### DISCUSSION TITLE
AppImage package support for Linux

### DISCUSSION THREAD

**Discussion by ayr-ton** [zed-industries/zed#25004] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Following the road to Linux (https://github.com/zed-industries/zed/issues/7015), I would ask you to investigate AppImage.

## What is AppImage

AppImage format is ideal for upstream packaging, which means that you get the software directly from the original author(s) without any intermediaries, exactly in the way the author(s) intended. And quickly. 

The experience is similar to Mac OS .dmg packages, so it's learning curve is smaller for Mac OS app developers. There's also apps like https://github.com/TheAssassin/AppImageLauncher that create seamless integration with all Linux distributions.

My favorite examples are: https://appimage.github.io/Todoist/ and  https://k8slens.dev/

## Competitors by learning curve

- AUR (for Arch Linux based) - easier, fast to get running and available to everyone: https://aur.archlinux.org/packages/zed-preview-bin
- Flatpak - Safer, some learning curve. Default for most distributions including Steamdeck and read-only OSes.
- Snap - Similar to flatpak, but with bigger learning curve for non cli like apps. By default is not in all linux distributions expect Ubuntu although has broad compatibility with all distros. 
- Official packages (one per distro: .deb, .rpm, etc) - more complex, an extra learning curve for each package type

## Getting started

- https://docs.appimage.org/introduction/index.html
- https://docs.appimage.org/packaging-guide/index.html
- https://docs.appimage.org/packaging-guide/optional/updates.html

**Comment by awedpotato767** [comment]
Issue #6687 comments on AppImages when mentioning flatpak alternatives. 
While I agree that the experience for AppImages may be familiar with macOS and some windows developers, most app centers on linux now have support for flatpaks (GNOME software and discover on KDE) out of the box.
AppImage is also getting quite dated technology-wise and the lack of automatic updates out of the box may cause headaches from users reporting bugs in out-of-date versions of the software.

**Comment by ayr-ton** [comment]
About the out-of-date versions issue:
https://docs.appimage.org/packaging-guide/optional/updates.html
TODOist and k8slens are also examples of self-updated functionality

**Comment by probonopd** [comment]
The AppImage format is about one app = one file simplicity. I'd say it is timeless, not "dated".  It allows you to keep multiple versions in parallel, and easily copy/move applications around between disks and computers. There is a mechanism for built-in updates. It is something entirely different that "installed" applications. More like PortableApps for Windows.

**Comment by BelKai** [comment]
I've been using official AppImages for Blender and Neovim ... , in combination with [gearlever](https://github.com/mijorus/gearlever). The experience has been amazing.
I'm really hoping the team will consider adding support for AppImage.

**Comment by lavilao** [comment]
adding appimage support would make changing computers/distros/containers quite easily as you can have everything in one usb.

**Comment by awedpotato767** [comment]
@lavilao I agree that they are convenient for some use cases, however surely the same distro neutrality is also provided by flatpaks? 
The only use case I can see for usb transfer is slow internet speeds/metered connections (which is valid).
While appImages are handy (I am not against them), I just think that they will probably be more effort than it's worth in terms of bad support ticket generation and upkeep cost, since zed is relatively new. 
Looking at blender and neovim's implementations of appimage packaging may be interesting, but I think having a single fully working build for everyone (on flatpak, the current standard) ought to be a priority over two "half baked" ones.
(shoutout to the contributors, zed on linux is amazing)

**Comment by lavilao** [comment]
@awedpotato767 to me the main beneffit is the portability, flatpaks while have the option to export to an offline "installer" the last time I tried it it only had the app package without the runtimes and it would not update from flathub.org because it was set to a local repository as source so they are not as portable (maybe this has changed but this was my experience).

**Comment by Lonniebiz** [comment]
AppImages are not "more trouble than their worth". All you have to do is put the AppImage creation into your build workflow. After that, its created automatically on each build.

I like AppImage way better than Flatpak and Snap, because when coupled with [firejail](https://firejail.wordpress.com/), you have way more control over what the application can access.

I currently use [Pulsar](https://pulsar-edit.dev/) for code editing, which provides an AppImage (that I can control with firejail). I'll continue using Pulsar until Zed releases an official AppImage each build. If that never happens, then I'll just stay with Pulsar.

**Comment by lavilao** [comment]
umm... for anyone interested I made an appimage of the latest zed pre-release. The repo is a mess but the appimage works. [here](https://github.com/lavilao/Zed-AppImage/releases) is the link.

**Comment by richardstevenhack** [comment]
Well, I tried using the flatpak.  Didn't work on openSUSE Tumbleweed, I don't know why. It simply wouldn't execute.

There was a version available for Tumbleweed - it no longer exists.

The reality is that programs installed by scripts are going to be a dying breed - because the large Linux distros are moving towards "immutable" which means no installation of anything except by snap, flatpak or AppImage.

So I vote for an AppImage maintained by someone who knows what they're doing - if there is any such person in open-source.

**Reply by debuggerdragon311** [reply]
i did...here  [zedx-appimage](https://github.com/debuggerdragon311/zedx-appimage.git)

**Comment by crnisamuraj** [comment]
Is it really that hard making github action that will build appimage with release? i can volunteer

**Comment by debuggerdragon311** [comment]
I built an AppImage version here: [zed-appimage](https://github.com/debuggerdragon311/zed-appimage.git)
Any feedback is helpful.

**Comment by richardstevenhack** [comment]
Needless to say, for security reasons only an AppImage supported or authorized by the developers can be trusted.

**Comment by Lonniebiz** [comment]
@richardstevenhack : And, as proof of that, I reported two bugs to the [AppImage project](https://github.com/debuggerdragon311/zedx-appimage.git) debuggerdraon311 provided and he closed both of them saying that the fact that it didn't work for my version of Linux was by design. So, I won't be using that offering.

**Reply by debuggerdragon311** [reply]
You can try now I did solved that issue. With a compatible version of the appimage. Any feedback is helpful.

**Comment by richardstevenhack** [comment]
Wonderful. AppImages are supposed to work on any Linux, just like FlatPaks. Depends on the implementer whether that's true, apparently. The wonders of open source...

I think I've decided not to consider Zed. I'll focus on vs-codium at the moment. With all the vibecoding tools available, who cares what the editor is as long as it supports them? The AI generates the code, you just have to examine and fix it - which doesn't require a whole lot of editing.  :-)
